File: dev_ws/src/learning_tf2_cpp/src/frame_listener.cpp
/*
 * @Description: 
 * @version: v1.0
 * @Author: HTY
 * @Date: 2022-07-26 22:53:51
 */

#include "rclcpp/rclcpp.hpp"
#include "tf2/exceptions.h"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "turtlesim/srv/spawn.hpp"

#include "geometry_msgs/msg/transform_stamped.hpp"
#include "geometry_msgs/msg/twist.hpp"


class FrameListener : public rclcpp::Node
{
public:
    explicit FrameListener(const std::string & node)
    : Node(node), turtle_spawn_service_ready(false), turtle_spawned(false)
    {
        this->declare_parameter<std::string>("source_frame", "turtle1");
        this->get_parameter("source_frame", source_frame);
        this->declare_parameter<std::string>("target_frame", "turtle2");
        this->get_parameter("target_frame", target_frame);

        tf_buffer = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);

        client = this->create_client<turtlesim::srv::Spawn>("spawn");
        pub = this->create_publisher<geometry_msgs::msg::Twist>("turtle2/cmd_vel", 10);
        timer = this->create_wall_timer(std::chrono::seconds(1), std::bind(&FrameListener::timer_callback, this));
    }

private:
    bool turtle_spawn_service_ready;
    bool turtle_spawned;
    std::string target_frame;
    std::string source_frame;
    std::unique_ptr<tf2_ros::Buffer> tf_buffer;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener;
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr client;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub;
    rclcpp::TimerBase::SharedPtr timer;

    void timer_callback()
    {
        if (turtle_spawn_service_ready)
        {
            if (turtle_spawned)
            {
                geometry_msgs::msg::TransformStamped trans;
                try
                {
                    trans = tf_buffer->lookupTransform(
                            target_frame,
                            source_frame,
                            tf2::TimePointZero
                            );
                }
                catch (tf2::TransformException & ex)
                {
                    RCLCPP_ERROR(this->get_logger(), "Could not transform %s to %s: %s",
                                 source_frame.c_str(),
                                 target_frame.c_str(),
                                 ex.what());
                    return;
                }

                geometry_msgs::msg::Twist msg;
                static const double scale_rotation_rate = 1.0;
                msg.angular.z = scale_rotation_rate * atan2(trans.transform.translation.y, trans.transform.translation.x);
                static const double scale_forward_rate = 0.5;
                msg.linear.x = scale_forward_rate * sqrt(pow(trans.transform.translation.y, 2) +
                                                         pow(trans.transform.translation.x, 2));
                pub->publish(msg);
            }
            else
            {
                RCLCPP_INFO(this->get_logger(), "Successfully spawned");
                turtle_spawned = true;
            }
        }
        else
        {
            if (client->service_is_ready())
            {
                auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
                request->x = 4.0;
                request->y = 2.0;
                request->theta = 0.0;
                request->name = target_frame;

                using ServiceResponseFuture = rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture;
                auto response_received_callback = [this](ServiceResponseFuture future)
                {
                    auto & result = future.get();
                    if (strcmp(result->name.c_str(), "turtle2") == 0)
                    {
                        turtle_spawn_service_ready = true;
                    }
                    else
                    {
                        RCLCPP_ERROR(this->get_logger(), "Service callback result mismatch");
                    }
                };
                auto result = client->async_send_request(request, response_received_callback);
            }
            else    // service ready
            {
                RCLCPP_ERROR(this->get_logger(), "Service is not ready");
            }
        }
    }
};


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<FrameListener>("frame_listener_node"));
    rclcpp::shutdown();
    return 0;
}

File: dev_ws/src/learning_tf2_python/learning_tf2_python/frame_listener.py
"""
 @Description: 
 @version: v1.0
 @Author: HTY
 @Date: 2022-07-26 16:02:25
"""

import math

import rclpy
from rclpy import time, duration
from rclpy.node import Node
from tf2_ros import TransformException, LookupException, ConnectivityException, ExtrapolationException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener

from turtlesim.srv import Spawn
from geometry_msgs.msg import Twist


class FrameListener(Node):
    def __init__(self, node):
        super(FrameListener, self).__init__(node)

        self.declare_parameter('source_frame', 'turtle1')
        self.source_frame = self.get_parameter('source_frame').get_parameter_value().string_value
        self.declare_parameter('target_frame', 'turtle2')
        self.target_frame = self.get_parameter('target_frame').get_parameter_value().string_value

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.spawner = self.create_client(Spawn, 'spawn')
        self.turtle_spawn_service_ready = False
        self.turtle_spawned = False
        self.result = None

        self.publisher = self.create_publisher(Twist, f'{self.target_frame}/cmd_vel', 10)

        self.timer = self.create_timer(1, self.timer_callback)

    def timer_callback(self):
        if self.turtle_spawn_service_ready:
            if self.turtle_spawned:
                try:
                    trans = self.tf_buffer.lookup_transform(
                        self.target_frame,
                        self.source_frame,
                        rclpy.time.Time(),
                        timeout=rclpy.duration.Duration(seconds=0.05)
                    )
                except TransformException as ex:
                    self.get_logger().error(f'Could not transform {self.target_frame} to {self.source_frame}: {ex}')
                    return
                except (LookupException, ConnectivityException, ExtrapolationException) as ex:
                    self.get_logger().error(f'Transform not ready: {ex}')
                    return

                msg = Twist()
                scale_rotation_rate = 1.0
                msg.angular.z = scale_rotation_rate * math.atan2(trans.transform.translation.y,
                                                                 trans.transform.translation.x)

                scale_forward_rate = 0.5
                msg.linear.x = scale_forward_rate * math.hypot(trans.transform.translation.x,
                                                               trans.transform.translation.y)

                self.publisher.publish(msg)

            else:   # turtle spawned
                if self.result.done():
                    self.get_logger().info(f'Successfully spawned {self.result.result().name}')
                    self.turtle_spawned = True
                else:
                    self.get_logger().warn('Spawn is not finished')

        else:   # service ready
            if self.spawner.service_is_ready():
                request = Spawn.Request()
                request.name = self.target_frame
                request.x = float(4)
                request.y = float(2)
                request.theta = float(0)

                self.result = self.spawner.call_async(request)
                self.turtle_spawn_service_ready = True
            else:
                # Check if the service is ready
                self.get_logger().info('Service is not ready')


def main(args=None):
    rclpy.init(args=args)
    listener = FrameListener('tf2_listener_node')
    rclpy.spin(listener)
    listener.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
